src/perf/multi_reader.cpp

```diff
@@ -0,0 +1,89 @@
+/*
+ * This file is part of the lo2s software.
+ * Linux OTF2 sampling
+ *
+ * Copyright (c) 2017,
+ *    Technische Universitaet Dresden, Germany
+ *
+ * lo2s is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * lo2s is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with lo2s.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <lo2s/perf/bio/reader.hpp>
+#include <lo2s/perf/bio/writer.hpp>
+#include <lo2s/perf/multi_reader.hpp>
+
+namespace lo2s
+{
+namespace perf
+{
+template <class Reader, class Writer>
+MultiReader<Reader, Writer>::MultiReader(trace::Trace& trace) : writer_(trace)
+{
+}
+
+template <class Reader, class Writer>
+int MultiReader<Reader, Writer>::addReader(ReaderIdentity identity)
+{
+    auto reader = readers_.emplace(std::piecewise_construct, std::forward_as_tuple(identity),
+                                   std::forward_as_tuple(Reader(identity)));
+    return reader.first->second.fd();
+}
+
+template <class Reader, class Writer>
+void MultiReader<Reader, Writer>::read()
+{
+    std::priority_queue<ReaderState, std::vector<ReaderState>, std::greater<ReaderState>>
+        earliest_available_;
+
+    for (auto& reader : readers_)
+    {
+        if (!reader.second.empty())
+        {
+            earliest_available_.push(ReaderState(reader.second.top()->time, reader.first));
+        }
+    }
+
+    while (!earliest_available_.empty())
+    {
+        auto state = earliest_available_.top();
+        earliest_available_.pop();
+
+        while (!readers_.at(state.identity).empty())
+        {
+
+            auto event = std::move(*readers_.at(state.identity).top());
```

Comment:
I don't see the point of dereferencing the pointer here. If it's needed, please explain in a comment.

Reply:
After the `readers_.at(state.identity).pop()` a few lines down, the event could be overwritten by the kernel ring buffer, so that why I copy it here.

However, I don't like copying here. So I have changed to logic so that pop() is only called after the event has been processed.